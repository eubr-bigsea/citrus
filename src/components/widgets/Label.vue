<template>
    <span class="label">
        <span :data-order="field.order" :data-name="field.name"
            :class="{'text-danger': error, required: field.required}" :title="error">{{field.label || field.name}}</span>
        <span v-show="field.required">*</span>:
        <span v-if="(showHelp === undefined || showHelp) && field.help" class="data-help">
            <span class="fa fa-question-circle float-right" :title="field.help"></span>
        </span>
    </span>
</template>

<script>
    import Widget from '../../mixins/Widget.js';
    export default {
        mixins: [Widget],
        name: "property-label",
        props: {
            field: { required: true, type: Object },
            value: {},
            showHelp: true,
        },
        computed: {
            error() {
                if (this.field.required && (this.value === undefined || this.value === null || this.value === '' || this.value === {})) {
                    return this.$tc('errors.missingRequiredValue');
                }
                return null;
            },
        }
    }
</script>
<style scoped>
    .label {
        font-size: 1em;
    }
</style>
