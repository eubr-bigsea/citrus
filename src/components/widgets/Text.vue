<template>
    <div v-if="readOnly">
        {{value === null ? field.default: value}}
    </div>
    <div v-else>
        <LabelComponent :field="field"
                        :value="value"
                        :show-help="showHelp" />
        <input type="text"
               maxlength="100"
               class="form-control input-sm"
               :value="value === null ? field.default: value"
               :required="field.required"
               :class="{'form-control-sm': small}"
               @input="updated">
    </div>
</template>
<script>
import LabelComponent from './Label.vue';
import Widget from '../../mixins/Widget.js';
import { debounce } from '../../util.js';
export default {
    name: 'TextComponent',
    components: { LabelComponent },
    mixins: [Widget],
    mounted() {
        const value = (this.field['default'] ? this.field['default']: null);
        this.$root.$emit(this.message,
            this.field, this.value || value);
    },
    methods: {
        updated: debounce(function (e) {
            this.triggerUpdateEvent(this.message, this.field, e.target.value);
        }, 500)
    },
};
</script>
