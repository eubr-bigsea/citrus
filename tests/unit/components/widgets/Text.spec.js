import { shallowMount } from '@vue/test-utils';
import Text from '$SRC/components/widgets/Text.vue';
import Label from '$SRC/components/widgets/Label.vue';
import { expect } from 'vitest';

describe('Text.vue', () => {
    afterEach(() => {
        vi.restoreAllMocks();
        vi.useRealTimers();
    });
    beforeEach(() => {
        vi.useFakeTimers({ toFake: ['setTimeout', 'clearTimeout', 'Date'] });
    });
    it('renders read-only widget with default value', () => {
        const wrapper = shallowMount(Text, {
            propsData: {
                field: {
                    value: null,
                    default: 'User input'
                },
                readOnly: true,
                value: null,
            }
        });
        expect(wrapper.findComponent(Label).exists()).toBe(false);
        expect(wrapper.text()).toBe("User input");
    });
    it('renders with field value', () => {
        const wrapper = shallowMount(Text, {
            propsData: {
                field: {
                    value: 3001,
                    default: 0
                },
                readOnly: false,
                value: null
            }
        });
        expect(wrapper.findComponent(Label).exists()).toBeTruthy();
        const inputText = wrapper.find('input[type="text"]');
        expect(inputText.exists()).toBeTruthy();

    });
    it('renders and emits after input ', async () => {
        const spy = vi.spyOn(Text.methods, 'updated');
        const wrapper = shallowMount(Text, {
            propsData: {
                field: {
                    value: 3001, default: 0
                },
                readOnly: false, value: null
            }
        });

        const inputText = wrapper.find('input');
        inputText.setValue(5000);
        vi.advanceTimersByTime(1000);
        const payload = wrapper.emitted();
        expect(payload.update[0][1]).toBe('5000');
        expect(spy).toBeCalled();

    });
    it('snapshots the component', () => {
        const wrapper = shallowMount(Text, { propsData: { field: {} } });
        expect(wrapper.text()).toMatchSnapshot();
    });
});
