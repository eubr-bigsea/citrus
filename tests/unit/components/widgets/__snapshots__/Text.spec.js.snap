// Vitest Snapshot v1

exports[`Integer.vue > snapshots the component 1`] = `""`;

exports[`Text.vue > snapshots the component 1`] = `""`;
